### PR TITLE
Add category additional description

### LINKED ISSUE #16627: [SEO | IMPROVEMENT] Add category bottom description
**Problem**
It would be very useful if you guys could add a second description field to a category, which would show up under the product list.

For SEO purposes, merchants usually add a lot of text to category description. Upper description is not very good place for this, because people will usually not read it and it just occupies space.

Collapse is one solution, but hidden text has less "value" for the robots.

**Solution**
Make a "copy" of category description and show it under a product list.

**Inspiration - my implementation**
Výstřižek

### PR DESCRIPTION
**Pull request by Hlavtox**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | This PR adds bottom category description field. This field is good for adding content for SEO purposes.
| Type?             | new feature
| Category?         | FO
| BC breaks?        | No - Only point I see is EditableCategory, I added the parameter at the end and provided a default value.
| Deprecations?     | no
| Fixed ticket?     | Fixes #16627
| Related PRs | https://github.com/PrestaShop/autoupgrade/pull/446
| How to test?      | Try to fill additional description in some categories and go check them displayed in FO.
| Possible impacts? | Should not be. If people do not update their template, they won't see it, no other side effects.

**Related theme PR**
https://github.com/PrestaShop/classic-theme/pull/18

**Screenshots**
BO
FO

Specs:https://github.com/PrestaShop/prestashop-specs/pull/305

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27292)

### CODE REVIEW

**Review by NeOMakinG — DISMISSED** [review]

File: classes/Category.php
Comment (kpodemski) [inline classes/Category.php]:
```suggestion
    public $additional_description;
```

By default, it would be at the bottom, but maybe we could have a more generic naming, ideas:
`secondary_description`
`additional_description`
`extra_description`

**Review by PululuK — COMMENTED** [review]
@Hlavtox 

Can you please update : 
- Db struct : file https://github.com/PrestaShop/PrestaShop/blob/develop/install-dev/data/db_structure.sql#L364
-  uupgrade sql scripts

**Comment by Hlavtox** [comment]
@PululuK It was somehow excluded from the changed files, should be fixed now. :-)

File: src/Core/Domain/Category/QueryResult/EditableCategory.php
Comment (PululuK) [inline src/Core/Domain/Category/QueryResult/EditableCategory.php]: Add BC label in this case, or set is optional param

Comment (PululuK) [reply src/Core/Domain/Category/QueryResult/EditableCategory.php]: Look good to me @Hlavtox

**Review by matks — CHANGES_REQUESTED** [review]
Integration tests fails so maybe a behavior was broken

**Comment by Hlavtox** [comment]
@matks Hi Mathieu, tests are fixed and I added the new field to tests. 👍
@kpodemski Whatever you think is the best. 😉

**Comment by kpodemski** [comment]
@PrestaShop/prestashop-maintainers 

Guys, what do you think about the field name? I'm more in favor of something more generic. Just like in the CSS, we don't have `section-left`, `section-bottom`, I can imagine someone using this field to have a description in, for example, the left or right sidebar of the page.

File: src/Core/Domain/Category/QueryResult/EditableCategory.php
Comment (sowbiba) [inline src/Core/Domain/Category/QueryResult/EditableCategory.php]: bottom_description vs description_bottom
it scratches me to have the words inverted in method names (same for localizedBottomDescription)

**Review by sowbiba — COMMENTED** [review]
Just a naming feedback
I also agree with @kpodemski `extra_description` or `additional_description` sound better for me

File: classes/Category.php
Comment (kpodemski) [reply classes/Category.php]: @Hlavtox let's go with `additional_description`

Comment (Hlavtox) [reply classes/Category.php]: @kpodemski Ok Krystian! :-)

**Comment by Hlavtox** [comment]
@kpodemski @sowbiba So the function nomenclature resolved itself to getAdditionalDescription. 😎

File: themes/classic/templates/catalog/_partials/category-footer.tpl
Comment (kpodemski) [inline themes/classic/templates/catalog/_partials/category-footer.tpl]: @Hlavtox this description shouldn't be displayed on the 2nd page, right?

Comment (Hlavtox) [reply themes/classic/templates/catalog/_partials/category-footer.tpl]: @kpodemski No it should be displayed all the time, the same as the upper content. Why hide it?

Comment (kpodemski) [reply themes/classic/templates/catalog/_partials/category-footer.tpl]: @Hlavtox to avoid duplicate content, upper description is hidden on the second page by default

https://github.com/PrestaShop/PrestaShop/blob/db334ee2069e946be7703cbca4b1168681d5a86b/themes/classic/templates/catalog/_partials/category-header.tpl#L26

**Comment by Hlavtox** [comment]
@kpodemski @matks Description conditioned as the first one + rebased

**Review by atomiix — DISMISSED** [review]

File: tests/Integration/Behaviour/Features/Context/Domain/CategoryFeatureContext.php
Comment (matks) [inline tests/Integration/Behaviour/Features/Context/Domain/CategoryFeatureContext.php]: I think here it should be camelCase

**Review by matks — CHANGES_REQUESTED** [review]
Hi @Hlavtox one very small issue and then it's good for me 👍  thanks for this great contribution.

As this has quite an impact I would be happy to have one more opinion, like @matthieu-rolland or @Matt75 to make sure this new addition is complete

Also don't forget @PululuK request:
> upgrade sql scripts

We need that shops that do upgrade from 1.7.x to v8 have this new SQL field added to database.

**Comment by Hlavtox** [comment]
@matks Fixed
@PululuK Already done ;-) https://github.com/PrestaShop/autoupgrade/pull/446

**Review by matks — DISMISSED** [review]

**Review by kpodemski — DISMISSED** [review]

**Comment by matks** [comment]
@Julievrz can you review wording so we can then move to QA 😊 ?

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/Blocks/form.html.twig
Comment (Julievrz) [inline src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/Blocks/form.html.twig]:
```suggestion
        'label': 'Additional description'|trans({}, 'Admin.Catalog.Feature'),
```

**Review by Julievrz — DISMISSED** [review]
Wording ✔️

**Review by kpodemski — DISMISSED** [review]

**Comment by Hlavtox** [comment]
@matks @atomiix @NeOMakinG bump pls

**Review by NeOMakinG — DISMISSED** [review]

**Comment by NeOMakinG** [comment]
Need Specs on the issue... @MatShir

**Review by atomiix — DISMISSED** [review]

**Review by khouloudbelguith — DISMISSED** [review]
Hi @Hlavtox,

I have some issues following these steps:

**First issue**

1. Go to BO > Catalog > Categories > Edit Home page
2. Click on save
2. See error

image

**Second issue**
1. Go to BO > Catalog > Categories > Add new category
2. In the "Additional description" field add `invalid characters: <>;=#{}`
3. Save
4. See error: no alert displayed 

https://user-images.githubusercontent.com/16067358/154484229-e0515372-25ba-40f0-818c-5d0837b05f37.mp4




**Third issue**
1. Go to BO > Catalog > Categories > Add new category
2. In the "Additional description" field add too long text > 21844 characters
3. Save
4. See error: no alert displayed

https://user-images.githubusercontent.com/16067358/154484243-e9e50d9a-85f5-4b14-9417-10e801f90379.mp4


Thanks!

**Comment by MatShir** [comment]
I'm wondering for the second and third issue if the alerts/errors messages are displayed for the other same fields as well. From my quick tests, they aren't. It might be a global issue.
Could you please confirm it ? If it is true then, the issue is not from the PR but a lack from the core that we could fix in another issue.

**Comment by Hlavtox** [comment]
Hello @khouloudbelguith

**1** - I just tested it and it works perfectly fine for me. See video - https://drive.google.com/file/d/1D2Kx7cEI7aGK6A8P9N4Q-qGdFNN4a15W/view?usp=sharing. However, I improved the wording of error conditions, so when you retest it, we can at least see if regular or lang field is causing the problem.

errors

**2+3** - As @MatShir stated (Thanks 🍺), this is an original behavior, we can fix it in another PR for both descriptions. 👍

**Comment by Hlavtox** [comment]
So, we tested with @khouloudbelguith and the error was related to linkrewrite being wrong, not this PR. All good to be merged. 👍

We agreed that I will open another PR that will implement ability to import this field from CSV, after this is merged.

**Review by atomiix — DISMISSED** [review]

**Review by kpodemski — DISMISSED** [review]

File: themes/classic/templates/catalog/_partials/category-footer.tpl
Comment (Progi1984) [inline themes/classic/templates/catalog/_partials/category-footer.tpl]:
```suggestion
{if $category.additional_description && $listing.pagination.items_shown_from == 1}
    <div id="js-product-list-footer">
        <div class="card">
            <div class="card-block">
                {$category.additional_description nofilter}
            </div>
        </div>
    </div>
{/if}
```

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/Blocks/form.html.twig
Comment (Progi1984) [inline src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/Blocks/form.html.twig:79]:
```suggestion
      {{ ps.form_group_row(categoryForm.additional_description, {}, {}) }}
```

File: src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php
Comment (Progi1984) [inline src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:125]:
```suggestion
                'label' => $this->trans('Additional description', 'Admin.Catalog.Feature'),
                'help' => $this->trans('Invalid characters', 'Admin.Notifications.Info') . ' <>;=#{}',
                'type' => FormattedTextareaType::class,
```

File: src/Core/Domain/Category/QueryResult/EditableCategory.php
Comment (Progi1984) [inline src/Core/Domain/Category/QueryResult/EditableCategory.php]:
```suggestion
    public function getAdditionalDescription(): array
```

File: src/Core/Domain/Category/Command/EditRootCategoryCommand.php
Comment (Progi1984) [inline src/Core/Domain/Category/Command/EditRootCategoryCommand.php]:
```suggestion
    public function setLocalizedAdditionalDescriptions(array $localizedAdditionalDescriptions): self
```

File: src/Core/Domain/Category/Command/EditRootCategoryCommand.php
Comment (Progi1984) [inline src/Core/Domain/Category/Command/EditRootCategoryCommand.php]:
```suggestion
    public function getLocalizedAdditionalDescriptions(): ?array
```

File: src/Core/Domain/Category/Command/EditCategoryCommand.php
Comment (Progi1984) [inline src/Core/Domain/Category/Command/EditCategoryCommand.php]:
```suggestion
    public function setLocalizedAdditionalDescriptions(array $localizedAdditionalDescriptions): self
```

File: src/Core/Domain/Category/Command/EditCategoryCommand.php
Comment (Progi1984) [inline src/Core/Domain/Category/Command/EditCategoryCommand.php]:
```suggestion
    public function getLocalizedAdditionalDescriptions(): ?array
```

File: src/Core/Domain/Category/Command/AddRootCategoryCommand.php
Comment (Progi1984) [inline src/Core/Domain/Category/Command/AddRootCategoryCommand.php]:
```suggestion
    public function setLocalizedAdditionalDescriptions(array $localizedAdditionalDescriptions): self
```

File: src/Core/Domain/Category/Command/AddRootCategoryCommand.php
Comment (Progi1984) [inline src/Core/Domain/Category/Command/AddRootCategoryCommand.php]:
```suggestion
    public function getLocalizedAdditionalDescriptions(): ?array
```

File: src/Core/Domain/Category/Command/AddCategoryCommand.php
Comment (Progi1984) [inline src/Core/Domain/Category/Command/AddCategoryCommand.php]:
```suggestion
    public function setLocalizedAdditionalDescriptions(array $localizedAdditionalDescriptions): self
```

File: src/Core/Domain/Category/Command/AddCategoryCommand.php
Comment (Progi1984) [inline src/Core/Domain/Category/Command/AddCategoryCommand.php]:
```suggestion
    public function getLocalizedAdditionalDescriptions(): ?array
```

**Review by Progi1984 — COMMENTED** [review]
@Hlavtox Thank you for your contribution. I have brought you some comments to your PR.

File: themes/classic/templates/catalog/_partials/category-footer.tpl
Comment (Hlavtox) [reply themes/classic/templates/catalog/_partials/category-footer.tpl]: @Progi1984 This is correct, the div has nothing to do with the description itself. Let people target it if they want. ;-)

**Review by kpodemski — DISMISSED** [review]

**Review by atomiix — DISMISSED** [review]

**Review by matks — DISMISSED** [review]

**Review by NeOMakinG — DISMISSED** [review]

**Review by khouloudbelguith — DISMISSED** [review]
Hi @Hlavtox,

I have an issue following these steps:

1. Fresh install of the PR with English language & France country
2. Go to BO > Catalog > Categories > Edit a category

image

Thanks!

**Comment by Hlavtox** [comment]
@khouloudbelguith I reverted the suggestion, now it will work, can you retest pls so we can finally merge?

File: themes/classic/templates/catalog/_partials/category-footer.tpl
Comment (Progi1984) [reply themes/classic/templates/catalog/_partials/category-footer.tpl]: @Hlavtox This is not fixed.

**Review by Progi1984 — COMMENTED** [review]
@Hlavtox Miss a comment. And it's good.

**Comment by Hlavtox** [comment]
@Progi1984 I don't understand what do you mean by `Miss a comment. And it's good.`.

I put back the previous way to enter a label and help, your suggestion did not work for some reason. 🤔

**Comment by matks** [comment]
Following the [move](https://github.com/PrestaShop/PrestaShop/pull/27592) of Classic theme to its [own repository](https://github.com/prestashop/classic-theme) you need to open a Pull Request that does modify the theme on the theme repository I think

**Comment by Hlavtox** [comment]
Rebased, made theme PR, made autoupgrade PR.

File: src/Core/Domain/Category/Command/AddCategoryCommand.php
Comment (atomiix) [inline src/Core/Domain/Category/Command/AddCategoryCommand.php]:
```suggestion
     * @var string[]|null
```

File: src/Core/Domain/Category/Command/AddRootCategoryCommand.php
Comment (atomiix) [inline src/Core/Domain/Category/Command/AddRootCategoryCommand.php]:
```suggestion
     * @var string[]|null
```

File: src/Core/Domain/Category/Command/EditCategoryCommand.php
Comment (atomiix) [inline src/Core/Domain/Category/Command/EditCategoryCommand.php]:
```suggestion
     * @var string[]|null
```

File: src/Core/Domain/Category/Command/AddCategoryCommand.php
Comment (atomiix) [inline src/Core/Domain/Category/Command/AddCategoryCommand.php]:
```suggestion
     * @return string[]|null
```

File: src/Core/Domain/Category/Command/AddRootCategoryCommand.php
Comment (atomiix) [inline src/Core/Domain/Category/Command/AddRootCategoryCommand.php]:
```suggestion
     * @return string[]|null
```

File: src/Core/Domain/Category/Command/EditCategoryCommand.php
Comment (atomiix) [inline src/Core/Domain/Category/Command/EditCategoryCommand.php]:
```suggestion
     * @return string[]|null
```

File: src/Core/Domain/Category/Command/EditRootCategoryCommand.php
Comment (atomiix) [inline src/Core/Domain/Category/Command/EditRootCategoryCommand.php]:
```suggestion
     * @var string[]|null
```

File: src/Core/Domain/Category/Command/EditRootCategoryCommand.php
Comment (atomiix) [inline src/Core/Domain/Category/Command/EditRootCategoryCommand.php]:
```suggestion
     * @return string[]|null
```

**Review by atomiix — APPROVED** [review]

**Review by kpodemski — APPROVED** [review]

**Review by khouloudbelguith — COMMENTED** [review]
Hi @Hlavtox,

In the FO, the additional description is not displayed.
I attached a screen record

https://user-images.githubusercontent.com/16067358/161042663-64c7c7bb-8995-4f44-b691-68317c1a9f4b.mp4

Thanks to check and feedback.

**Comment by kpodemski** [comment]
Hello @khouloudbelguith 

I think you need to update the classic theme with the related PR from the description. Unfortunately, with the classic theme being in the dedicated repository testing is a bit more complicated.

**Comment by Hlavtox** [comment]
Thanks @kpodemski 👍

@khouloudbelguith You must also install the related theme PR - https://github.com/PrestaShop/classic-theme/pull/18.

**Comment by Hlavtox** [comment]
@khouloudbelguith Can you please test it while you have the PR installed so it does not have to wait another two months for QA? It could have been merged since 8.2.2022….

**Review by Progi1984 — APPROVED** [review]

**Review by khouloudbelguith — COMMENTED** [review]
Hi @Hlavtox,

It is ok :heavy_check_mark: 

- automatic tests => ok
[Mochawesome Report_category.pdf](https://github.com/PrestaShop/PrestaShop/files/8425145/Mochawesome.Report_category.pdf)
- Import categories via csv => ok
- Add a category via api (with additional category) => ok
- Add a root category (multistore enabled)
- Add a category from the product page => ok
- Add/Edit a category in different shops => ok
- Add/Edit category with disabled language / deleted language / RTL lanaguage / New lanaguage created manually

https://user-images.githubusercontent.com/16067358/161936602-54d0bebe-b476-4e9e-a076-11a98a19ccc6.mp4


https://user-images.githubusercontent.com/16067358/161936641-f7414ffb-c394-4a7f-a0a0-3da1ddce93a0.mp4

Thanks!

**Comment by kpodemski** [comment]
🎉 

thank you @Hlavtox @khouloudbelguith

**Comment by matks** [comment]
Hoora @Hlavtox 😄

**Pull request merged**
